compiler/rustc_hir_analysis/src/check/wfcheck.rs

```diff
@@ -1750,13 +1765,53 @@ fn check_method_receiver<'tcx>(
             _ =>
             // Report error; would not have worked with `arbitrary_self_types[_pointers]`.
             {
-                tcx.dcx().emit_err(errors::InvalidReceiverTy { span, receiver_ty })
+                match receiver_validity_err {
+                    ReceiverValidityError::DoesNotDeref => {
+                        tcx.dcx().emit_err(errors::InvalidReceiverTy { span, receiver_ty })
+                    }
+                    ReceiverValidityError::MethodGenericParamUsed => {
+                        tcx.dcx().emit_err(errors::InvalidGenericReceiverTy { span, receiver_ty })
+                    }
+                }
             }
         });
     }
     Ok(())
 }
 
+/// Error cases which may be returned from `receiver_is_valid`. These error
+/// cases are generated in this function as they may be unearthed as we explore
+/// the `autoderef` chain, but they're converted to diagnostics in the caller.
+enum ReceiverValidityError {
+    /// The self type does not get to the receiver type by following the
+    /// autoderef chain.
+    DoesNotDeref,
+    /// A type was found which is a method type parameter, and that's not allowed.
+    MethodGenericParamUsed,
+}
+
+/// Confirms that a type is not a type parameter referring to one of the
+/// method's type params.
+fn confirm_type_is_not_a_method_generic_param(
+    ty: Ty<'_>,
+    method_generics: Option<&hir::Generics<'_>>,
+) -> Result<(), ReceiverValidityError> {
+    if let ty::Param(param) = ty.kind() {
+        if let Some(generics) = method_generics {
+            if generics
+                .params
+                .iter()
+                .filter(|g| matches!(g.kind, GenericParamKind::Type { .. }))
+                .map(|g| g.name.ident().name)
+                .contains(&param.name)
```

Comment:
@compiler-errors is checking if the `ty::Param` name matches the name of a generic parameter on the method the right way to do this check?

Reply:
Definitely not lol, that is not hygienic and also unnecessarily iterating through all the generics.

Reply:
Just compare the index of the param to the `parent_count`: https://doc.rust-lang.org/stable/nightly-rustc/rustc_middle/ty/generics/struct.Generics.html

Reply:
I had a crack at this in c2cbf1d8848d9bacb4a3035646405a57b5f90e73.